src/lib.rs

```diff
@@ -134,49 +160,48 @@ impl IoUringBufRing {
     /// # Safety
     ///
     /// caller must make sure release [`IoUringBufRing`] with correct `ring`
-    pub unsafe fn release<S, C>(self, ring: &IoUring<S, C>) -> io::Result<()>
+    pub unsafe fn release<S, C>(mut self, ring: &IoUring<S, C>) -> io::Result<()>
     where
         S: squeue::EntryMarker,
         C: cqueue::EntryMarker,
     {
         ring.submitter().unregister_buf_ring(self.buf_group)?;
 
+        ManuallyDrop::drop(&mut self.buf_ring_mmap);
+        ManuallyDrop::drop(&mut self.bufs);
         Ok(())
     }
 
-    fn init_bufs_with_iter<B: ExactSizeIterator<Item = Vec<u8>>>(&mut self, bufs: B) {
-        let ring_entries = bufs.len();
-        for (id, mut buf) in bufs.enumerate() {
+    fn init_bufs_with_iter(&mut self) {
+        let mask = self.mask();
+        let mmap = self.buf_ring_mmap.get_mut();
+        let bufs = self.bufs.get_mut();
+        for (id, buf) in bufs.iter_mut().enumerate() {
             // Safety: all arguments are valid
             unsafe {
-                self.add_buffer(
-                    buf.as_mut(),
-                    buf.capacity(),
-                    id as _,
-                    (ring_entries - 1) as _,
+                mmap.add_buffer(
+                    // The buffers are created with `with_capacity`.
+                    buf.spare_capacity_mut(),
                     id as _,
+                    mask,
+                    id,
                 );
             }
-
-            self.bufs.get_mut().push(buf);
         }
 
-        self.advance_buffers(ring_entries as _);
+        unsafe {
+            mmap.advance_buffers(bufs.len() as _);
+        }
     }
 
     /// # Safety
     ///
-    /// caller must make sure release valid buffer
-    unsafe fn release_borrowed_buffer(&self, buffer: &mut BorrowedBuffer) {
-        self.add_buffer(
-            buffer.buf.as_mut(),
-            buffer.buf.capacity(),
-            buffer.id,
-            self.mask(),
-            0,
-        );
-
-        self.advance_buffers(1);
+    /// * Caller must make sure release valid buffer
+    /// * Callers should be on the same thread.
```

Comment:
> on the same thread

I think this is unnecessary:

1. the `IoUringBufRing` is not `Sync` or `Send` (maybe we can make it `Send` in the future)
2. `IoUring` is `Send`, there is no problem to send the ring and the buf-ring to another thread

Reply:
Oh yes. Of course callers are on the same thread. It is necessary to avoid data race.